cips/cip-tx-limits.md

```diff
@@ -0,0 +1,34 @@
+| cip | TBD(28) |
+| - | - |
+| title | Transaction size limit |
+| description | Set limit for transaction size |
+| author | Josh Stein ([@jcstein](https://github.com/jcstein)), Nina Barbakadze ([@ninabarbakadze](https://github.com/ninabarbakadze)), Rootul Patel ([@rootulp](https://github.com/rootulp)) |
+| discussions-to | <https://forum.celestia.org/t/cip-limit-number-of-pfbs-and-non-pfbs-per-block-increase-transaction-size-limit/1843> |
+| status | Draft |
+| type | Standards Track |
+| category | Core |
+| created | 2024-10-16 |
+
+## Abstract
+
+This CIP proposes to set the limit for transaction size. The proposal is to set the transaction size limit to 2MiB. Setting the transaction size limit is consensus-breaking.
+
+## Specification
+
+1. Transaction size is limited to 2MiB by setting the versioned parameter `MaxTxSize` to 2097152, which is 2MiB in bytes. From version v3 and above, in `CheckTx`, `PrepareProposal`, and `ProcessProposal`, each transaction's size is checked against the `appconsts.MaxTxSize` threshold. This ensures that transactions over the limit are rejected or excluded at all stages, from initial submission to execution.
+
+## Rationale
+
+The rationale for this proposal is to set the transaction size limit to 2MiB, even with 8MiB blocks, to prevent issues with gossiping large transactions. Gossiping an 8MiB transaction without chunking could be detrimental to the network. This is a consensus-breaking change.
```

Comment:
@ninabarbakadze can you elaborate here on why we implemented this as a consensus breaking change instead of modifying the default mempool `MaxTxBytes` to 2 MiB?

I asked about it in a few places (Slack) and https://github.com/celestiaorg/celestia-app/issues/3686#issuecomment-2393738597

Reply:
I'll wait on feedback until tomorrow and plan to get draft in then. Thank you @ninabarbakadze and @rootulp !

Reply:
because every node can set the default to any value they choose to and afaik we don't know the effect large transactions are going to have on the network so in my understanding it's preventative. cc @evan-forbes

Reply:
Agreed the motivation for the change is preventative.

> because every node can set the default to any value they choose to

Nodes can override the default but I expect most nodes will use the default. Assuming most nodes use the default config, transactions larger than 2 MiB won't be included in blocks so blob submitters will continue submitting blobs <= 2 MiB so that their blobs get included in blocks.

Reply:
I agree with you but it seemed like we wanted to have a hard limit on it.

Reply:
Yea, I'm wondering why did we want the hard limit? The non-consensus breaking change is still preventative.

Reply:
Suggested revision below: https://github.com/celestiaorg/CIPs/pull/226#discussion_r1806974380